Framework/API/inc/MantidAPI/TriviallyParallelAlgorithm.h

```diff
@@ -0,0 +1,56 @@
+#ifndef MANTID_API_TRIVIALLYPARALLELALGORITHM_H_
+#define MANTID_API_TRIVIALLYPARALLELALGORITHM_H_
+
+#include "MantidAPI/Algorithm.h"
+#include "MantidAPI/DllConfig.h"
+
+namespace Mantid {
+namespace API {
+
+/** Base class for algorithms that treat all spectra independently, i.e., we can
+  trivially parallelize over the spectra without changes. The assumption is that
+  we have one input and one output workspace. The storage mode is just
+  propagated from input to output. When a specific algorithm is determined to be
+  trivially parallel (this is a manual process), the only required change to add
+  MPI support is to inherit from this class instead of Algorithm. Inheriting
+  from TriviallyParallelAlgorithm instead of from Algorithm provides the
+  necessary overriden method(s) to allow running an algorithm with MPI. This
+  works under the following conditions:
+  1. The algorithm must have a single input and a single output workspace.
+  2. No output files may be written since filenames would clash.
+
+  @author Simon Heybrock
+  @date 2017
+
+  Copyright &copy; 2017 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_API_DLL TriviallyParallelAlgorithm : public Algorithm {
```

Comment:
The idea of this class is great, but I am not a big fan of the name. Would Easy or Simple be better than Trivially? I know Trivially has the right connotations when talking about parallelisation, but it is a bit cumbersome.

Reply:
As you say, `TriviallyParallel` probably has a commonly understood meaning, while I would not know what `EasyParallel` or `SimpleParallel` would imply. How about `EmbarrassinglyParallelAlgorithm`?

Reply:
For me `PerfectlyParallelAlgorithm` makes a statement about the performance, computational complexity, or scaling behavior of the algorithm. In general algorithms that would use the base class are *not* scaling perfectly, in some cases not at all.

As an example, `LoadInstrument` is loading the *complete* instrument on each MPI rank. There is absolutely no speed up, but the implementation is trivial -- there are no changes or pieces of code that deal with the fact that there may be many other ranks running the same algorithm for the "same" workspace.

This is why I prefer `TriviallyParallelAlgorithm`, or something similar. I do however completely understand your criticism -- I also think it is a bit unwieldy and, while imho better than the other alternatives, also not entirely clear. How about `ImplicitlyParallelAlgorithm`?

@martyngigg Can we have your opinion as an "outsider" on this naming question?

Reply:
I see the concerns here. I think it's important that the name convey something about the implementation requirements for the developer and I do like that `EmbarrassinglyParallel` is a more commonly understood term with regards to parallelization so that's what would get my vote.

Reply:
Reading https://en.wikipedia.org/wiki/Embarrassingly_parallel, it also seems to imply that the problem is easily split into sub-tasks. Are people fine with that, even if it is not correct for cases like `LoadInstrument`?

Reply:
Another option would be to simply go with `ParallelAlgorithm`. That does not imply much in particular and details can be found in the documentation.

Reply:
This would be fine with me.

Reply:
:+1: